### DISCUSSION TITLE
set_track_fields in addition to set_fields

### DISCUSSION THREAD

**Discussion by bertbesser** [beetbox/beets#3925] [Ideas]
Hi there and thx 4 beets,

my goal is to add certain fields _at track level_ automatically during import. E.g. a comment like 'managed by the magnificent beets tool'. 😅 Those fields should be written to files.

There is a similar feature, the `set_fields` importer option and the accompanying `--set` CLI option. Hoever, these work at album level. I would like to implement `set_track_fields` resp. `--set-track` with analogous semantics at the track level. From quite some discussions out on the internet, I noticed that this is not only my wish, but also the desire of other people.

Are there concerns/objections/blockers for the additional `set_track_fields` or does it have a good chance to be merged? Asking b/c I'd rather save the precious time if the feature is not welcome.

Example config and CLI call:
```
...
import:
  set_fields:
    comments: 'managed by the magnificent beets tool'
    ...
  set_track_fields:
    comments: 'managed by the magnificent beets tool'
    ...
...
```
```
%> beet import --set-track=comments='managed by the magnificent beets tool' ...
```

The suggestion to implement an additional option `set_track_fields` instead of adding a flag to `set_fields` to switch between 'album mode' and 'track mode' is for two reasons:
- backwards compatibility and
- fine grained control over the track-level and album-level fields of each track.

Let me know what you think.

Regards and thx again,
bb

**Comment by sampsyo** [comment]
Hi there! This does make sense, but I want to point out a little bit of backstory: this has come up in https://github.com/beetbox/beets/issues/1881#issuecomment-808552290, at which point @wisp3rwind's follow-up pointed to an alternative: https://github.com/beetbox/beets/pull/2988 can obviate the need for some cases where you want to set item-level attributes on all of an album's tracks.

I think that doesn't exactly work for your use case because you actually want to set on-disk (e.g., ID3) tags—you don't want these for queries.

I think we probably should do something about this, but I'm not certain about the `set_track_fields` design. The main reason is that `set_fields` *does* already set track fields—it only does it in singleton import mode. So maybe we want to just automatically set all track-level fields on albums in album mode, or maybe we want that behavior to be behind a (boolean) flag.

**Reply by bertbesser** [reply]
Hi,

yes, I checked the difference between singleton and album mode of `set_fields`, ie. that they set fields on the track-level resp. album-level.

I was also thinking about a boolean flag to decide whether fields are set on just one level or on both album-level and track-level. For backwards compatibility the flag cannot be a child of `set_fields` and hence would be a sibling of `set_fields`. That would imply some cross-refs in the docs and possibly some unclear config files out there (where `set_fields` and the flag are not next to each other).

Cheers

**Reply by bertbesser** [reply]
> So maybe we want to just automatically set all track-level fields on albums in album mode.

This would actually be my favorite: Just set fields on both album and all tracks by default automatically. My only fear is that this might break some users' setups.

**Reply by sampsyo** [reply]
It's worth a shot, certainly! Care to open a pull request?

**Reply by bertbesser** [reply]
Sure. I think I will find some time over the weekend. Will ref from here.

**Reply by bertbesser** [reply]
Found some time earlier. Please see #3927.
